### PR TITLE
Add Quick Start for M2M application

### PR DESCRIPTION
**Pull request by asha15**
Screenshot from 2024-03-28 15-09-11

### Related Issues
- Issue `#1` or (None)

### Related PRs
- Related PR `#1` or (None)

### Checklist
- [ ] e2e cypress tests locally verified.
- [ ] Manual test round performed and verified.
- [ ] UX/UI review done on the final implementation.
- [ ] Documentation provided. (Add links if there are any)
- [ ] Unit tests provided. (Add links if there are any)
- [ ] Integration tests provided. (Add links if there are any)

### Security checks
- [ ] Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines?
- [ ] Ran FindSecurityBugs plugin and verified report?
- [ ] Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets?

### CODE REVIEW

File: apps/console/src/extensions/application-templates/shared/components/m2m-custom-configuration.tsx [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+                                    <CopyInputField
+                                        value={ copyScopesValue }
+                                        data-componentid={ `${ componentId }-scope-readonly-input` }
+                                    />
+                                </Form.Field>
+                                <Form.Field>
+                                    <label>
+                                        {
+                                              t("extensions:console.application.quickStart" +
+                                              ".m2m.customConfig.tokenRequest")
```
Comment (pavinduLakshan) [inline apps/console/src/extensions/application-templates/shared/components/m2m-custom-configuration.tsx:299]:
```suggestion
                                          t("extensions:console.application.quickStart" +
                                             ".m2m.customConfig.tokenRequest")
```

File: apps/console/src/extensions/application-templates/shared/components/m2m-custom-configuration.tsx [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+            <Card.Content textAlign="left">
+                <Text muted>
+                    <Trans
+                        i18nKey={
+                            "extensions:develop.applications.quickstart" +
+                            ".m2mApp.configurations.heading"
+                        }
+                    >
+                        A Machine-to-Machine Application represents a program that interacts with an API without any user involvement. 
+                        Execute a{" "}
```
Comment (pavinduLakshan) [inline apps/console/src/extensions/application-templates/shared/components/m2m-custom-configuration.tsx:346]: 'Execute a..' looks odd IMO. Have we got the text content reviewed by the doc team?

File: apps/console/src/extensions/application-templates/shared/components/m2m-custom-configuration.tsx [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+        inboundProtocolConfig,
+        ["data-componentid"]: componentId
+    } = props;
+
+    const { t } = useTranslation();
+
+    const { getLink } = useDocumentation();
+
+    const path: string[] = history.location.pathname.split("/");
+    const appId: string = path[path.length - 1].split("#")[0];
```
[the remark this answers is not in the document]
Comment (asha15) [reply apps/console/src/extensions/application-templates/shared/components/m2m-custom-configuration.tsx]: Yes. please check https://github.com/wso2/identity-apps/blob/8eb0c3fe565d0ff1ba1d38ef9a06ecb840f74153/features/admin-applications-v1/components/api-authorization/api-authorization.tsx#L103

File: apps/console/src/extensions/application-templates/shared/components/m2m-custom-configuration.tsx [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+    useEffect(() => {
+        if (allAuthorizedScopes) {
+            setCopyScopesValue(allAuthorizedScopes.map(
+                (scope: AuthorizedPermissionListItemInterface) => scope.name).join(" ")
+            );
+        }
+    }, [ allAuthorizedScopes ]);
+
+    const onAPIAuthorizationTabClick = (): void => {
+        onTriggerTabUpdate(APIAuthorizationIndex);
```
[the remark this answers is not in the document]
Comment (asha15) [reply apps/console/src/extensions/application-templates/shared/components/m2m-custom-configuration.tsx]: fixed with https://github.com/wso2/identity-apps/pull/5850/commits/f0cfb4daf5d26c9d34f01e01e87e48aafb33701f

File: apps/console/src/extensions/application-templates/shared/components/m2m-custom-configuration.tsx [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+                                                ".configurations.scope.label")
+                                        }
+                                        <Hint className="mt-0 mb-0" popup>
+                                            <Trans
+                                                i18nKey={
+                                                    "extensions:develop.applications.quickstart" +
+                                                    ".m2m.configurations.scopes"
+                                                }
+                                            >
+                                                These are the set of scopes that are authorized.
```
[the remark this answers is not in the document]
Comment (asha15) [reply apps/console/src/extensions/application-templates/shared/components/m2m-custom-configuration.tsx]: Fixed with https://github.com/wso2/identity-apps/pull/5850/commits/43e15da9364840776d4ac2ddcf767bee4b07232b

File: apps/console/src/extensions/application-templates/shared/components/m2m-custom-configuration.tsx (hunk 1 above)
Comment (asha15) [reply apps/console/src/extensions/application-templates/shared/components/m2m-custom-configuration.tsx]: Fixed with https://github.com/wso2/identity-apps/pull/5850/commits/2ddd84863170374c76df3baa232a575f90d495f1

File: apps/console/src/extensions/application-templates/shared/components/m2m-custom-configuration.tsx [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+                                        showLineNumbers={ false }
+                                        language="shell"
+                                        options={ {
+                                            lineWrapping: true
+                                        } }
+                                        height="100%"
+                                        theme="dark"
+                                        sourceCode={
+                                            tokenRequest
+                                        }
```
[the remark this answers is not in the document]
Comment (asha15) [reply apps/console/src/extensions/application-templates/shared/components/m2m-custom-configuration.tsx]: Fixed with https://github.com/wso2/identity-apps/pull/5850/commits/2ddd84863170374c76df3baa232a575f90d495f1

File: apps/console/src/extensions/application-templates/templates/m2m-application/m2m-application.json [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    "description": "Applications developed to target m2m applciation.",
+    "enabled": true,
+    "displayOrder": 5,
+    "image": "oidcMobile",
+    "name": "M2M Application",
+    "subTemplatesSectionTitle": "Technology",
+    "types": [
+        "android",
+        "ios",
+        "flutter"
```
[the remark this answers is not in the document]
Comment (asha15) [reply apps/console/src/extensions/application-templates/templates/m2m-application/m2m-application.json]: Fixed with https://github.com/wso2/identity-apps/pull/5850/commits/528b79ec08d1ebc472bc6decd4f942c2a1a4ed68

File: apps/console/src/extensions/i18n/resources/en-US/extensions.ts [hunk 8]
```diff
@@ -1295,6 +1301,17 @@ export const extensions: Extensions = {
                             "application that should not be secured, and do not require authentication. Multiple " +
                             "URIs can be set using <3>comma separated</3> values."
                     }
+                },
+                m2m: {
+                    configurations: {
+                        clientSecret: {
+                            hint: "The OAuth 2.0 Client Secret valid at the authorization server."
```
[the remark this answers is not in the document]
Comment (asha15) [reply apps/console/src/extensions/i18n/resources/en-US/extensions.ts:1331]: @divyaamunugama Could you please check this.

File: apps/console/src/extensions/application-templates/shared/components/m2m-custom-configuration.tsx (hunk 2 above)
Comment (asha15) [reply apps/console/src/extensions/application-templates/shared/components/m2m-custom-configuration.tsx]: Adding @divyaamunugama

**Comment by divyaamunugama** [comment]
### UI string suggestions:

**_Description_**
```
A Machine-to-Machine Application operates independently, interacting with an API without requiring user input. It uses the client credentials grant flow to obtain an access token for authorized API access.
```

**_Hint texts for fields:_**

- Client ID: `The Client ID generated when registering the application in Asgardeo.`
- Client Secret: `The Client Secret generated when registering the application in Asgardeo.`
- Token Endpoint: `The token endpoint of Asgardeo.`
- Scope: `The list of OIDC scopes that are used by the application for requesting information from Asgardeo.`
- Token Request: `The API request used by the application to obtain an access token from Asgardeo.`

File: apps/console/src/extensions/i18n/resources/en-US/extensions.ts [hunk 9]
```diff
@@ -1317,6 +1324,17 @@ export const extensions: Extensions = {
...
+                        },
+                        tokenRequest: "The API request used by the application to obtain an access token from Asgardeo.",
+                        tokenEndpoint: "The token endpoint of Asgardeo.",
+                        scopes: "The list of scopes that are authorized." +
```
Comment (pavinduLakshan) [inline apps/console/src/extensions/i18n/resources/en-US/extensions.ts:1335]:
```suggestion
                        scopes: "The list of authorized scopes." +
```

File: apps/console/src/extensions/i18n/resources/en-US/extensions.ts [hunk 10]
```diff
@@ -1317,6 +1324,17 @@ export const extensions: Extensions = {
...
+                                "Please add more scopes by authorizing APIs from the <1>API Authorization</1> tab."
```
Comment (pavinduLakshan) [inline apps/console/src/extensions/i18n/resources/en-US/extensions.ts:1336]: We don't normally use the word `Please` in UI text.

File: apps/console/src/extensions/i18n/resources/en-US/extensions.ts (hunk 9 above)
Comment (asha15) [reply apps/console/src/extensions/i18n/resources/en-US/extensions.ts]: Fixed with https://github.com/wso2/identity-apps/pull/5850/commits/8ea62a2b2fbb0e869c091fbde5ce85f5ef91d8b5

File: apps/console/src/extensions/i18n/resources/en-US/extensions.ts [hunk 11]
```diff
@@ -1317,6 +1324,20 @@ export const extensions: Extensions = {
                             "application that should not be secured, and do not require authentication. Multiple " +
                             "URIs can be set using <3>comma separated</3> values."
                     }
+                },
+                m2m: {
+                    configurations: {
+                        clientId: {
+                            hint: "The Client ID generated when registering the application in Asgardeo."
```
Comment (pavinduLakshan) [inline apps/console/src/extensions/i18n/resources/en-US/extensions.ts:1331]: Lets define a placeholder for the product name and pass the product name through i18n options.

Please check and fix the other places as well.

```suggestion
                            hint: "The Client ID generated when registering the application in {prpductName}."
```

File: apps/console/src/extensions/i18n/resources/en-US/extensions.ts [hunk 12]
```diff
@@ -1317,6 +1324,20 @@ export const extensions: Extensions = {
...
+                        },
+                        clientSecret: {
+                            hint: "The Client Secret generated when registering the application in Asgardeo."
+                        },
+                        tokenRequest: "The API request used by the application to obtain an access token from Asgardeo.",
+                        tokenEndpoint: "The token endpoint of Asgardeo.",
+                        scopes: "The list of authorized scopes." +
+                                "Include required scopes by authorizing APIs through the <1>API Authorization</1> tab"
```
Comment (pavinduLakshan) [inline apps/console/src/extensions/i18n/resources/en-US/extensions.ts:1339]: WDYT?

```suggestion
                                "Include additional required scopes by authorizing APIs through the <1>API Authorization</1> tab"
```

File: features/admin.extensions.v1/application-templates/templates/m2m-application/__tests__/quick-start.test.tsx [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+        inboundProtocolConfig: {},
+        onApplicationUpdate: jest.fn(),
+        onTriggerTabUpdate: jest.fn(),
+        template: {
+            id: "test-template-id",
+            name: "M2MApplication"
+        }
+    };
+
+    it("renders the M2MApplicationQuickStart component", () => {
```
Comment (pavinduLakshan) [inline features/admin.extensions.v1/application-templates/templates/m2m-application/__tests__/quick-start.test.tsx:39]: If you are able to run this unit test spec, shall we add another test to verify the token request command updating with selected scopes?

File: features/admin.extensions.v1/application-templates/templates/m2m-application/__tests__/quick-start.test.tsx [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+        render(<M2MApplicationQuickStart { ...defaultProps } />, { });
+
+        const M2Mguide: Element = screen.getByTestId("m2m-app-quick-start");
```
Comment (pavinduLakshan) [inline features/admin.extensions.v1/application-templates/templates/m2m-application/__tests__/quick-start.test.tsx:42]:
```suggestion
        const m2mGuide: Element = screen.getByTestId("m2m-app-quick-start");
```

File: features/admin.extensions.v1/application-templates/templates/m2m-application/quick-start.tsx [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+/**
+ * M2M App Quick start content.
+ *
+ * @param  props - Props injected into the component.
+ * @returns M2M App QuickStart.
+ */
+const M2MApplicationQuickStart:
+    FunctionComponent<M2MApplicationQuickStartPropsInterface> = (
+        props:M2MApplicationQuickStartPropsInterface
+    ): ReactElement => {
```
Comment (pavinduLakshan) [inline features/admin.extensions.v1/application-templates/templates/m2m-application/quick-start.tsx:59]: Let's use ReactNode instead.

https://dev.to/fromaline/jsxelement-vs-reactelement-vs-reactnode-2mh2

**Review by pavinduLakshan — DISMISSED** [review]

File: features/admin.extensions.v1/application-templates/templates/m2m-application/__tests__/quick-start.test.tsx [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { render, screen } from "@testing-library/react";
```
Comment (pavinduLakshan) [inline features/admin.extensions.v1/application-templates/templates/m2m-application/__tests__/quick-start.test.tsx:19]: We expose a modified render method for our unit tests in the test utils file. Let's use that function here by importing from `features/test-configs/utils`.

```suggestion
import { render, screen } from "../../../test-configs/utils";
```

File: features/admin.extensions.v1/application-templates/templates/m2m-application/__tests__/quick-start.test.tsx [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+import React from "react";
+import "@testing-library/jest-dom/extend-expect";
```
Comment (pavinduLakshan) [inline features/admin.extensions.v1/application-templates/templates/m2m-application/__tests__/quick-start.test.tsx:21]: extend-expect is no longer being exposed from testing library.

```suggestion
```

File: apps/console/src/extensions/i18n/resources/en-US/extensions.ts (hunk 12 above)
Comment (asha15) [reply apps/console/src/extensions/i18n/resources/en-US/extensions.ts:1339]: IMO it's better to proceed without the `additional` word as sometimes there will be no authorized scopes for the newly created application.

File: apps/console/src/extensions/i18n/resources/en-US/extensions.ts (hunk 11 above)
Comment (asha15) [reply apps/console/src/extensions/i18n/resources/en-US/extensions.ts]: Fixed with https://github.com/wso2/identity-apps/pull/5850/commits/3b40c638a941e99c8405c5a1650041f3e872d3c5

File: features/admin.extensions.v1/application-templates/templates/m2m-application/__tests__/quick-start.test.tsx (hunk 14 above)
Comment (asha15) [reply features/admin.extensions.v1/application-templates/templates/m2m-application/__tests__/quick-start.test.tsx]: Fixed with: https://github.com/wso2/identity-apps/pull/5850/commits/3b40c638a941e99c8405c5a1650041f3e872d3c5

File: features/admin.extensions.v1/application-templates/templates/m2m-application/__tests__/quick-start.test.tsx (hunk 16 above)
Comment (asha15) [reply features/admin.extensions.v1/application-templates/templates/m2m-application/__tests__/quick-start.test.tsx]: Fixed with: https://github.com/wso2/identity-apps/pull/5850/commits/5a6848d86668e0e09927f2a01bcdd5c94638196b

File: features/admin.extensions.v1/application-templates/templates/m2m-application/__tests__/quick-start.test.tsx (hunk 17 above)
Comment (asha15) [reply features/admin.extensions.v1/application-templates/templates/m2m-application/__tests__/quick-start.test.tsx]: Fixed with: https://github.com/wso2/identity-apps/pull/5850/commits/5a6848d86668e0e09927f2a01bcdd5c94638196b

**Review by pavinduLakshan — APPROVED** [review]

**Pull request merged**